marlowe-playground-client/src/Simulation.purs

```diff
@@ -400,616 +276,429 @@ marloweActionInput isEnabled f current =
-            ]
-        ]
-    , tbody_ (flattenPayments payments)
-    ]
+  choicesData =
+    let
+      (choices :: Array _) = state ^. (_marloweState <<< _Head <<< _state <<< _choices <<< to Map.toUnfoldable)
 
-flattenPayments :: forall p. Array Payment -> Array (HTML p HAction)
-flattenPayments payments =
-  concatMap
-    ( \(Payment party (Assets mon)) ->
+      asTuple (Tuple (ChoiceId choiceName choiceOwner) value) = show choiceName /\ show choiceOwner /\ show value /\ mempty /\ mempty /\ unit
+    in
+      map asTuple choices
+
+  paymentsData =
+    let
+      payments = state ^. (_marloweState <<< _Head <<< _payments)
+
+      asTuple :: Payment -> Array (String /\ String /\ String /\ String /\ String /\ Unit)
+      asTuple (Payment party (Assets mon)) =
         concatMap
-          ( \(Tuple curr toks) ->
-              map
-                ( \(Tuple tok val) ->
-                    renderPayment party curr tok val
-                )
-                $ Map.toUnfoldable toks
+          ( \(Tuple currencySymbol tokenMap) ->
+              ( map
+                  ( \(Tuple tokenName value) ->
+                      (show party /\ currencySymbol /\ tokenName /\ show value /\ mempty /\ unit)
+                  )
+                  (Map.toUnfoldable tokenMap)
+              )
           )
-          $ Map.toUnfoldable mon
-    )
-    payments
-
-renderPayment :: forall p. Party -> CurrencySymbol -> TokenName -> BigInteger -> HTML p HAction
-renderPayment party currSym tokName money =
-  tr []
-    [ td_
-        [ text (show party)
-        ]
-    , td
-        [ class_ $ ClassName "middle-column"
-        ]
-        [ text (show currSym)
-        ]
-    , td
-        [ class_ $ ClassName "middle-column"
-        ]
-        [ text (show tokName)
-        ]
-    , td
-        [ class_ $ ClassName "left-border-column"
-        ]
-        [ text (show money)
-        ]
-    ]
+          (Map.toUnfoldable mon)
+    in
+      concatMap asTuple payments
+
+  bindingsData =
+    let
+      (bindings :: Array _) = state ^. (_marloweState <<< _Head <<< _state <<< _boundValues <<< to Map.toUnfoldable)
+
+      asTuple (Tuple (ValueId valueId) value) = show valueId /\ show value /\ mempty /\ mempty /\ mempty /\ unit
+    in
+      map asTuple bindings
+
+  tableRow title emptyMessage a b c d e [] = emptyRow title emptyMessage
+
+  tableRow title emptyMessage a b c d e rowData = headerRow title a b c d e <> foldMap (\dataTuple -> uncurry5 row dataTuple) rowData
+
+  headerRow title a b c d e =
+    [ div [ classes [ rTableCell, first, Classes.header ] ] [ text title ]
+    , div [ classes [ rTableCell, rTableCell, Classes.header ] ] [ text a ]
+    , div [ classes [ rTableCell, rTableCell, Classes.header ] ] [ text b ]
+    , div [ classes [ rTableCell, rTableCell, Classes.header ] ] [ text c ]
+    , div [ classes [ rTableCell, rTableCell, Classes.header ] ] [ text d ]
+    , div [ classes [ rTableCell, rTableCell, Classes.header ] ] [ text e ]
```

Comment:
And these should be a `map` over that list

Reply:
the list thing applies but I have put the parameters in a list now and mapped over it as it neatens things up a bit.